src/libraries/System.Security.Cryptography.Cose/src/System/Security/Cryptography/Cose/CoseMessage.cs

```diff
@@ -0,0 +1,184 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Formats.Cbor;
+using System.Runtime.Versioning;
+
+namespace System.Security.Cryptography.Cose
+{
+    public abstract class CoseMessage
+    {
+        private const byte EmptyStringByte = 0xa0;
+        // COSE tags https://datatracker.ietf.org/doc/html/rfc8152#page-8 Table 1.
+        internal const CborTag Sign1Tag = (CborTag)18;
+
+        internal byte[]? _content;
+        internal byte[] _signature;
+        internal byte[] _protectedHeaderAsBstr;
+
+        private CoseHeaderMap _protectedHeaders;
+        private CoseHeaderMap _unprotectedHeaders;
+        public CoseHeaderMap ProtectedHeader => _protectedHeaders;
+        public CoseHeaderMap UnprotectedHeader => _unprotectedHeaders;
+
+        internal CoseMessage(CoseHeaderMap protectedHeader, CoseHeaderMap unprotectedHeader, byte[]? content, byte[] signature, byte[] encodedProtectedHeader)
+        {
+            _content = content;
+            _signature = signature;
+            _protectedHeaderAsBstr = encodedProtectedHeader;
+            _protectedHeaders = protectedHeader;
+            _unprotectedHeaders = unprotectedHeader;
+        }
+
+        // Sign and MAC also refer to the content as payload.
+        // Encrypt also refers to the content as cyphertext.
+        public ReadOnlyMemory<byte>? Content
+        {
+            get
+            {
+                if (_content  != null)
+                {
+                    return _content;
+                }
+
+                return null;
+            }
+        }
+
+        public static CoseSign1Message DecodeSign1(byte[] cborPayload)
+        {
+            try
+            {
+                var reader = new CborReader(cborPayload);
+                CborTag? tag = DecodeTag(reader);
+                if (tag != null && tag != Sign1Tag)
+                {
+                    throw new CryptographicException(SR.Format(SR.DecodeSign1IncorrectTag, tag));
+                }
+
+                CoseSign1Message message = DecodeCoseSign1Core(reader);
+                return reader.BytesRemaining == 0 ? message : throw new CryptographicException(SR.Format(SR.DecodeSign1ErrorWhileDecoding, SR.DecodeSign1MesageContainedTrailingData));
```

Comment:
Is there a scenario where someone knows they have the start of a COSE_Sign1 and don't know how long it is?  If so, consider `DecodeSign1(ReadOnlySpan<byte> cborPayload, out int bytesRead)` which doesn't throw on an incomplete read.

Reply:
We don't have an scenario for that yet, we should probably hold off until then IMO.